mesonbuild/cmake/interpreter.py

```diff
@@ -185,8 +189,10 @@ def artifact(self, name: str) -> T.Optional[T.Union['ConverterTarget', 'Converte
             keys += [self._rel_artifact_key(i), os.path.basename(i), self._base_artifact_key(i)]
         return self._return_first_valid_key(keys)
 
-    def generated(self, name: str) -> T.Optional[T.Union['ConverterTarget', 'ConverterCustomTarget']]:
-        return self._return_first_valid_key([self._rel_generated_file_key(name), self._base_generated_file_key(name)])
+    def generated(self, name: str) -> T.Optional['ConverterCustomTarget']:
+        res = self._return_first_valid_key([self._rel_generated_file_key(name), self._base_generated_file_key(name)])
+        assert res is None or isinstance(res, ConverterCustomTarget)
```

Comment:
Is this assertion really needed here? Shouldn't `_return_first_valid_key` ensure that instead?

Reply:
Yes. `_return_first_valid_key` is just a fancy dict lookup (for a dict of type `T.Dict[str, T.Union['ConverterTarget', 'ConverterCustomTarget']]`) and does no further checking.

This assertion works because of the `add` logic, which ensures that only `ConverterCustomTarget` objects are stored for any keys generated in the `generated` function.

Reply:
So, in short, `_return_first_valid_key` can also return both types in other instances, but it would be a bug (and blow up further down the line anyway) if it were to return a `ConverterTarget` in the scope of `generated`.

I have to admit that this is not the best piece of software I have ever designed...

Reply:
ahhh, okay. that makes sense.